docs/source/migration.mdx

```diff
@@ -545,6 +566,52 @@ await startStandaloneServer(server, {
 
 </MultiCodeBlock>
 
+If you want to migrate quickly to Apollo Server 4 *without* altering your data sources, the snippet below replicates Apollo Server 3's `dataSources` behavior with a custom plugin.
+
+<ExpansionPanel title="Click to expand">
+
+```ts
+import { ApolloServerPlugin, BaseContext } from '@apollo/server';
+import type { DataSource } from 'apollo-datasource';
+
+type DataSources = Record<string, DataSource>;
+type DataSourcesFn<TDataSources extends DataSources = DataSources> =
+  () => TDataSources;
+
+interface ContextWithDataSources<TDataSources extends DataSources = DataSources>
+  extends BaseContext {
+  dataSources?: TDataSources;
+}
+
+export const ApolloDataSources = <
+  TDataSources extends DataSources = DataSources,
+>(options: {
+  dataSources: DataSourcesFn<TDataSources>;
+}): ApolloServerPlugin<ContextWithDataSources<TDataSources>> => ({
+  requestDidStart: async (requestContext) => {
+    const dataSources = options.dataSources();
+    const initializers = Object.values(dataSources).map(async (dataSource) => {
+      if (dataSource.initialize)
+        dataSource.initialize({
+          cache: requestContext.cache,
+          context: requestContext.contextValue,
+        });
+    });
+
+    await Promise.all(initializers);
+
+    requestContext.contextValue.dataSources = dataSources;
+  },
+});
+
+// 
+const plugins = [ApolloDataSources({ dataSources })];
+const server = buildServer({ schema, plugins });
```

Comment:
Not sure the `buildServer` abstraction is helpful here, it's just hiding how AS is constructed.

```suggestion
const server = new ApolloServer({
  schema,
  plugins: [ApolloDataSources({ dataSources })],
});
```